### PR TITLE
feat(sdk): Optimise how `LinkedChunk::insert_gap_at` works when inserting at first position

### PR DESCRIPTION
**Pull request by Hywan**
Imagine we have this linked chunk:

```rust
assert_items_eq!(linked_chunk, ['a'] [-] ['b', 'c'] ['d', 'e', 'f']);
```

Before the patch, when we were running:

```rust
let position_of_d = linked_chunk.item_position(|item| *item == 'd').unwrap();
linked_chunk.insert_gap_at((), position_of_d)?;
```

it was taking `['d', 'e', 'f']` to split it at index 0, so `[]` + `['d', 'e', 'f']`, and then was inserting a gap in between, thus resulting into:

```rust
assert_items_eq!(linked_chunk, ['a'] [-] ['b', 'c'] [] [-] ['d', 'e', 'f']);
```

The problem is that it creates a useless empty chunk. It's a waste of space, and rapidly, of computation. When used with `EventCache`, this problem occurs every time a backpagination occurs (because it executes a `replace_gap_at` to insert the new item, and then executes a `insert_gap_at` if the backpagination contains another `prev_batch` token).

With this patch, the result of the operation is now:

```rust
assert_items_eq!(linked_chunk, ['a'] [-] ['b', 'c'] [-] ['d', 'e', 'f']);
```

The `assert_items_eq!` macro has been updated to be able to assert empty chunks. The `test_insert_gap_at` has been updated to test all edge cases.

---

These commits are extracted from https://github.com/matrix-org/matrix-rust-sdk/pull/3230.

### CODE REVIEW

**Review by poljar — APPROVED** [review]
Sorry, took a bit longer since I'm unfamiliar with the code. Left some complaints but overall looks sensible.

File: crates/matrix-sdk/src/event_cache/linked_chunk.rs [hunk 1]
```diff
@@ -283,6 +286,11 @@ impl<Item, Gap, const CAP: usize> LinkedChunk<Item, Gap, CAP> {
                 }
             };
 
+            new_chunk_ptr = chunk
+                .next
+                // SAFETY: A new `Chunk` has just been inserted, so it exists.
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk.rs:316]: Sure.

File: crates/matrix-sdk/src/event_cache/linked_chunk.rs [hunk 2]
```diff
@@ -305,7 +313,10 @@ impl<Item, Gap, const CAP: usize> LinkedChunk<Item, Gap, CAP> {
         // pointer, which is valid and well aligned.
         let _chunk_boxed = unsafe { Box::from_raw(chunk_ptr.as_ptr()) };
 
-        Ok(())
+        Ok(
+            // SAFETY: `new_chunk_ptr` is valid, non-null and well-aligned.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk.rs:341]: Sir, yes Sir!

File: crates/matrix-sdk/src/event_cache/linked_chunk.rs [hunk 3]
```diff
@@ -197,6 +197,30 @@ impl<Item, Gap, const CAP: usize> LinkedChunk<Item, Gap, CAP> {
...
+        // the invariant: a `Gap` cannot be the first chunk.
+        if item_index == 0 && chunk.is_items() && chunk.previous.is_some() {
+            let previous_chunk = chunk
+                .previous_mut()
+                // SAFETY: The `previous` chunk exists because we have tested
+                // `chunk.previous.is_some()` in the `if` statement.
+                .unwrap();
+
+            previous_chunk.insert_next(Chunk::new_gap_leaked(
+                chunk_identifier_generator.generate_next().unwrap(),
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/linked_chunk.rs:214]: Done, https://github.com/matrix-org/matrix-rust-sdk/pull/3256
